cheroot/server.py

```diff
@@ -1968,14 +1977,41 @@ def bind_unix_socket(self, bind_addr):  # noqa: C901  # FIXME
         return sock
 
     @staticmethod
-    def prepare_socket(bind_addr, family, type, proto, nodelay, ssl_adapter):
+    def _make_socket_reusable(socket_):
+        # Most BSD kernels implement SO_REUSEPORT the way that only the
+        # latest listener can read from socket. Some of BSD kernels also
+        # have SO_REUSEPORT_LB that works similarly to SO_REUSEPORT
+        # in Linux.
+        if hasattr(socket, 'SO_REUSEPORT_LB'):
+            socket_.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT_LB, 1)
+        elif hasattr(socket, 'SO_REUSEPORT'):
+            socket_.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
+        elif IS_WINDOWS:
+            socket_.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        else:
+            raise NotImplementedError(
+                'Current platform does not support port reuse',
+            )
+
+    @classmethod
+    def prepare_socket(
+            cls, bind_addr, family, type, proto, nodelay, ssl_adapter,
+            reuse_port=False,
+    ):
         """Create and prepare the socket object."""
         sock = socket.socket(family, type, proto)
         connections.prevent_socket_inheritance(sock)
 
         host, port = bind_addr[:2]
         IS_EPHEMERAL_PORT = port == 0
 
+        if reuse_port:
+            if sock.family not in (socket.AF_INET, socket.AF_INET6):
+                raise ValueError('Cannot reuse a non-IP socket')
+            if IS_EPHEMERAL_PORT:
+                raise ValueError('Cannot reuse an ephemeral port (0)')
```

Comment:
I think I'd like to see this code moved to `_make_socket_reusable`, something like this:

```diff
diff --git a/cheroot/server.py b/cheroot/server.py
index 4ffeb69c..16fd9ac6 100644
--- a/cheroot/server.py
+++ b/cheroot/server.py
@@ -1977,7 +1977,16 @@ class HTTPServer:
         return sock
 
     @staticmethod
-    def _make_socket_reusable(socket_):
+    def _make_socket_reusable(socket_, bind_addr):
+        host, port = bind_addr[:2]
+        IS_EPHEMERAL_PORT = port == 0
+
+        if socket_.family not in (socket.AF_INET, socket.AF_INET6):
+            raise ValueError('Cannot reuse a non-IP socket')
+
+        if IS_EPHEMERAL_PORT:
+            raise ValueError('Cannot reuse an ephemeral port (0)')
+
         # Most BSD kernels implement SO_REUSEPORT the way that only the
         # latest listener can read from socket. Some of BSD kernels also
         # have SO_REUSEPORT_LB that works similarly to SO_REUSEPORT
@@ -2006,11 +2015,7 @@ class HTTPServer:
         IS_EPHEMERAL_PORT = port == 0
 
         if reuse_port:
-            if sock.family not in (socket.AF_INET, socket.AF_INET6):
-                raise ValueError('Cannot reuse a non-IP socket')
-            if IS_EPHEMERAL_PORT:
-                raise ValueError('Cannot reuse an ephemeral port (0)')
-            cls._make_socket_reusable(sock)
+            cls._make_socket_reusable(bind_addr, sock)
 
         if not (IS_WINDOWS or IS_EPHEMERAL_PORT):
             """Enable SO_REUSEADDR for the current socket.
```

I know there's some repetition around `IS_EPHEMERAL_PORT`, but that seems preferable to having the re-use port logic in two places.

WDYT?